### ISSUE TITLE
MIDI Control Change not recognized by LV2 plugins

### ISSUE THREAD

**Issue by polluxsynth** [zynthian/zynthian-issue-tracking#744]
MIDI Control Change messages (at least not sustain pedal (CC64) or mod wheel (CC1)) are not recognized by LV2 plugins, i.e.
operating the sustain pedal or mod wheel has no effect on LV2 plugins, but it works fine on at least some of the standalone synth engines, and on LV2 plugins at least on an older stable branch.

More specifically, I've verified that Obxd, Dexed and Pianoteq (when run as an LV2 plugin) don't respond to the sustain pedal (CC 64).
Whereas Pianoteq in the standalone version and ZynAddSubFX respond fine to sustain pedal messages.
I compared this with an old stable version I had lying around, where both Obxd and Dexed responded fine to
CC64 and CC1 (didn't test Pianoteq in stable).

**To Reproduce**
1. Add Obxd, Dexed or PIanoteq (LV2 plugin).
2. Note how none of them respond to the sustain pedal.
3. To verify that the pedal is actually working and getting through the Zynthian MIDI routing, set up a ZynAddSubFX layer and note how it responds fine to the sustain pedal.

**Expected behavior**
LV2 plugins should respond to CC64 and CC1 (within the limitations of the specific plugins, of course).

**Setup:**
 - SD-image 2021-09-25-zynthianos-stable.img
 - Hardware Config
  Raspberry Pi 4 Model B Rev 1.4
  Audio: [HifiBerry DAC+ ADC PRO]
  Display: [ZynScreen 3.5 (v1)]
  Wiring: [MCP23017_ZynScreen]
  I2C: [MCP23017@0x20, MCP23008@0x21, ADS1115@0x48, MCP4728@0x61]
 - Software Versions
 zyncoder: testing (6eb5496)
zynthian-ui: [testing (2bfb33a)]
zynthian-sys: [testing (af606bf)]
zynthian-data: [testing (f29c69d)]
zynthian-webconf: [zynthian-engine-mixer-rename (b23b0ae)]

(The zynthian-webconf branch is a fix allowing the webconf to start at all, there being a bug in the latest testing branch when combined with zynthian-ui. A pull request for this has been submitted).

**Additional context**
I'm guessing the problem lies with JALV or its configuration.

**Comment by riban-bw** [comment]
I confirm this is an issue most likely caused by recent change to default MIDI learn. CC are not passed to engines not mapped in its MIDI learn map. These engines do not have a specific LV2 port (control) for sustain and hence do not have a default mapping. I will consider how best to resolve this. We want a default set of CC that are not cleared by MIDI learn. (This is how it used to work but had some undesirable side effects.)

**Comment by riban-bw** [comment]
I tried to send the CC if it is not learned but this results in the CC being sent twice to MIDI output, once direct from MIDI input and a second time when MIDI learn filter resends it. This will take a bit more consideration...

**Comment by riban-bw** [comment]
I fixed this by sending MIDI CC to the engine if it was not processed by the MIDI learn mechanism, i.e. MIDI input goes to MIDI learn filter and if not configured is sent back to MIDI router where it only goes to the internal engine, not external MIDI output hence avoiding duplicate CC messages.

This means that an engine receives all CC and handles it natively unless the CC is mapped to a LV2 port (control) within Zynthian.

@jofe please validate this logic. I think it makes sense but may have side effects or changed behaviour because CC which may previously have been filtered may now reach engine (which is engine's native and hence expected behaviour) and it may be awkward to block such CC (although this should be possible with MIDI filter).

**Comment by polluxsynth** [comment]
Does this mean that, with the fix, if the MIDI learn filter is set up to map a given CC, let's say CC22, that CC64 (sustain pedal) will not be forwarded implicitly? Or is CC64 always forwarded to the plugin, unless CC64 is explicitly mapped in the learn filter?

**Comment by riban-bw** [comment]
All CC will be sent to the MIDI input unless the CC is mapped to another function. Mapping CC 21 to a control will stop CC 21 from being sent to the engine's MIDI input whilst other CC will still be sent to the engine.

**Comment by polluxsynth** [comment]
Good, that sounds excellent.

Thanks for the quick fix! I quickly verified last night with the PIanoteq LV2 that at least CC64 is now getting through as expected.

**Comment by jofemodo** [comment]
Jalv LV2 engine also keep a list of specific MIDI controllers for some plugins. You can check it here:

https://github.com/zynthian/zynthian-ui/blob/b69072ed9b5f00cd2601af71375099545d60eef7/zyngine/zynthian_engine_jalv.py#L119

As you can see, the list doesn't include MIDI controllers for Dexed, OBXd and Pianoteq, but it includes controllers for MDA DX10, MDA JX10, MDA ePiano, MDA Piano and reMID. Yes, it's a little bit outdated and it should be completed.

Perhaps we could improve this "feature" instead of overcomplicating things in ZynMidiRouter and MIDI-learning code. Also, i don't like the idea of not having the possibility of learning the hiden MIDI controllers.

Please, consider reverting this change and improving the per-plugin MIDI-controller list.

**Comment by riban-bw** [comment]
This patch has introduced an issue with setBfree which sends its control changes to zynmidirouter. This results in a howl-round of messages. We need to change this behaviour.

**Comment by jofemodo** [comment]
It's not only setBfree. It's a more general mechanism that could be used by other engines. BTW, the ZMIP_CTRL port was created for allowing this mechanism to work.
Revert the change totally, please. It's easy to revert and the issue can be solved in a different way. MIDI controllers should be available from the GUI, so you can control and learn them.

**Comment by jofemodo** [comment]
Anyway, we should think a better mechanism for keeping this list of per-plugin MIDI controllers:

https://github.com/zynthian/zynthian-ui/blob/b69072ed9b5f00cd2601af71375099545d60eef7/zyngine/zynthian_engine_jalv.py#L119

**Comment by riban-bw** [comment]
I reverted the changes in midi router and zyngui then used the mechanism described by @jofemodo to add an extra (default) page of MIDI controllers to jalv engines. By default, MIDI Synth LV2 will get a page with Modulation Wheel and Sustain Pedal controllers mapped to CC1 & CC64. These are normal CC bindings so can be removed or the CC may be mapped to a different control, e.g. PianoTeq exposes its own sustain pedal (which is now on page 2) so a user may prefer to map to that (although this may only be of benefit if the pedal is variable - on/off pedals will have the same behaviour whichever mapping is used). I started to check the LV2 synths and found most supported sustain pedal. Some don't and some don't handle CC1 by default but it seemed reasonable to leave these two as the default mapping. It is possible to override this default for individual instruments (hard coded) which I have done for Raffo Synth that supports neither hence I removed the default MIDI Controllers page from Raffo Synth. Other overrides that were previously there have not been reinstated because they were inhibiting working controllers, e.g. Obxd was configured to use neither but supports both.

So we should now have working sustain pedal and modulation wheel for all LV2 synths that support them. We can add overrides to hide unsupported controllers as we find them and also add extra controls as required, e.g. some LV2 support CC7 volume.

I did try to implement a mechanism for overriding existing LV2 controllers, e.g. assigning CC64 to Pianoteq's native sustain pedal but this proved too challenging and time consuming so I removed that and implemented this simple solution. It works!
